include/dlaf/cublas_tile.h

```diff
@@ -0,0 +1,200 @@
+    DLAF_CUBLAS_CALL(cublasZtrsm(std::forward<Args>(args)...));
+  }
+};
+
+template <typename T>
+struct CublasGemm;
+
+template <>
+struct CublasGemm<float> {
+  template <typename... Args>
+  static void call(Args&&... args) {
+    DLAF_CUBLAS_CALL(cublasSgemm(std::forward<Args>(args)...));
+  }
+};
+
+template <>
+struct CublasGemm<double> {
+  template <typename... Args>
+  static void call(Args&&... args) {
+    DLAF_CUBLAS_CALL(cublasDgemm(std::forward<Args>(args)...));
+  }
+};
+
+template <>
+struct CublasGemm<std::complex<float>> {
+  template <typename... Args>
+  static void call(Args&&... args) {
+    DLAF_CUBLAS_CALL(cublasCgemm(std::forward<Args>(args)...));
+  }
+};
+
+template <>
+struct CublasGemm<std::complex<double>> {
+  template <typename... Args>
+  static void call(Args&&... args) {
+    DLAF_CUBLAS_CALL(cublasZgemm(std::forward<Args>(args)...));
+  }
+};
+
+template <typename T>
+struct CublasHerk;
+
+template <>
+struct CublasHerk<std::complex<float>> {
+  template <typename... Args>
+  static void call(Args&&... args) {
+    DLAF_CUBLAS_CALL(cublasCherk(std::forward<Args>(args)...));
+  }
+};
+
+template <>
+struct CublasHerk<std::complex<double>> {
+  template <typename... Args>
+  static void call(Args&&... args) {
+    DLAF_CUBLAS_CALL(cublasZherk(std::forward<Args>(args)...));
+  }
+};
+
+}
+
+template <class T>
+void cublasTrsm(cublasHandle_t handle, const blas::Side side, const blas::Uplo uplo, const blas::Op op,
+                const blas::Diag diag, const T alpha, const matrix::Tile<const T, Device::GPU>& a,
+                const matrix::Tile<T, Device::GPU>& b) {
+  SizeType m = b.size().rows();
+  SizeType n = b.size().cols();
+
+  DLAF_ASSERT(a.size().rows() == a.size().cols(), "`a` is not square!", a);
+
+  auto left_side = (side == blas::Side::Left ? m : n);
+  DLAF_ASSERT(a.size().rows() == left_side, "`a` has an invalid size!", a, left_side);
+
+  internal::CublasTrsm<T>::call(handle, util::blasToCublas(side), util::blasToCublas(uplo),
+                                util::blasToCublas(op), util::blasToCublas(diag), m, n,
+                                util::blasToCublasCast(&alpha), util::blasToCublasCast(a.ptr()), a.ld(),
+                                util::blasToCublasCast(b.ptr()), b.ld());
+}
+
+template <class T>
+void cublasGemm(cublasHandle_t handle, const blas::Op op_a, const blas::Op op_b, const T alpha,
+                const matrix::Tile<const T, Device::GPU>& a, const matrix::Tile<const T, Device::GPU>& b,
+                const T beta, const matrix::Tile<T, Device::GPU>& c) {
+  SizeType m;
+  SizeType k;
+  if (op_a == blas::Op::NoTrans) {
+    m = a.size().rows();
+    k = a.size().cols();
+  }
+  else {
+    m = a.size().cols();
+    k = a.size().rows();
+  }
+  SizeType k2;
+  SizeType n;
+  if (op_b == blas::Op::NoTrans) {
+    k2 = b.size().rows();
+    n = b.size().cols();
+  }
+  else {
+    k2 = b.size().cols();
+    n = b.size().rows();
+  }
```

Comment:
Just a minor thought.

I'm wondering if in the future wouldn't be worth to factor out this kind of checks, that in general are not device-dependent.

Reply:
Good idea. Let me also follow up on this in a separate PR?

Reply:
Absolutely! 😉